### ISSUE TITLE
docker container remembers major/minor of a block device when restarted - can cause data corruption

### ISSUE THREAD

**Issue by alex-zadara** [docker/docker#26569]
**Description**
When creating a docker container, we can attach a block device to it, using, for example "--device /dev/mapper/mydevice:/v1:rw". At this point, the container will resolve the device name to "major:minor" and persist it. When the container is started, it will use the "major:minor" to access the block device. 

Now if we reboot the host machine, then the same "major:minor" can refer to a completely different block device. For example, if we had several Device Mapper objects (e.g., created by LVM) before the reboot, then after reboot they may be re-created in different order, thus receiving different "major"minor" numbers. But a container that was created before the reboot will use the "major:minor" it was created with. Now writing data to this "major:minor" from within the container can actually access an unrelated block device, completely corrupting its data.

**Steps to reproduce the issue:**
1. Create two device mappers manually:
dmsetup create vol1 --table "0 1024 linear /dev/sda 0"
dmsetup create vol2 --table "0 1024 linear /dev/sdb 0"
Make sure there are no other device mappers in the system.
2. Create a docker container, and attach vol1 to it using "--device /dev/mapper/vol1:/vol1:rw"
3. Run the container and write some data to /dev/vol1
4. Reboot the machine
5. Create the device mappers in different order:
dmsetup create vol2 --table "0 1024 linear /dev/sdb 0"
dmsetup create vol1 --table "0 1024 linear /dev/sda 0"
Now "/dev/mapper/vol2" should receive the same "major:minor" that "/dev/mapper/vol1" had before the reboot.
6. Run the same container
7. Now accessing "/dev/vol1" from within the container will actually write onto sdb, corrupting its data.

**Output of `docker version`:**
Client:
 Version:      1.9.1
 API version:  1.22
 Go version:   go1.5.1
 Git commit:   fefbb50-dirty
 Built:        Wed Sep 14 12:31:18 UTC 2016
 OS/Arch:      linux/amd64

Server:
 Version:      1.9.1
 API version:  1.22
 Go version:   go1.5.1
 Git commit:   fefbb50-dirty
 Built:        Wed Sep 14 12:31:18 UTC 2016
 OS/Arch:      linux/amd64

**Output of `docker info`:**
root@vsa-000000ad-vc-1:~# docker info
Containers: 1
Images: 2
Server Version: 1.9.1
Storage Driver: btrfs
Execution Driver: native-0.2
Logging Driver: json-file
Plugins: 
 Volume: local
 Network: host bridge null
Kernel Version: 3.18.19-zadara05
Operating System: Ubuntu 14.04.4 LTS (containerized)
OSType: linux
Architecture: x86_64
CPUs: 3
Total Memory: 6.076 GiB
Name: vsa-000000ad-vc-1
ID: SYQF:3HN6:63Q5:Q4XD:3UEV:IZBQ:4N27:KY2B:HQG6:OUF6:R6PU:2XW7
WARNING: No swap limit support

What would be the recommendation to deal with this issue? Can docker be fixed to persist the device name (like /dev/mapper/vol1 or /dev/sdb)? I understand that in some cases it still won't work, because after reboot /dev/sdb might refer to a different block device. But this will at least give the user an option to use persistent names, which she can arrange to always map to the same block device after reboot.

**Comment by alex-zadara** [comment]
Greetings, docker community. Any update on this issue?

**Comment by justincormack** [comment]
I think persisting the name might be a little better, although it may be
hard to do as the kernel only knows about the number.

**Comment by alex-zadara** [comment]
@justincormack 
Yes, the container should persist the name. And when the container starts, it should resolve the name to major:minor and continue as today. Key thing is that every time container should resolve the name to current major:minor.
